modules/core/src/main/java/org/apache/synapse/mediators/builtin/ForEachMediator.java

```diff
@@ -96,85 +105,140 @@ public boolean mediate(MessageContext synCtx) {
-        DetachedElementContainer detachedElementContainer =
-                getDetachedMatchingElements(processingEnvelope, synCtx, expression);
+            //Read the complete JSON payload from the synCtx
+            String jsonPayload = JsonUtil.jsonPayloadToString(((Axis2MessageContext) synCtx).getAxis2MessageContext());
+            DocumentContext parsedJsonPayload = JsonPath.parse(jsonPayload);
 
-        List<?> splitElements = detachedElementContainer.getDetachedElements();
+            // SynapseJSONPath implementation reads the JSON stream and execute the JSON path.
+            JsonElement iterableChildElements = parsedJsonPayload.read(((SynapseJsonPath) expression).getJsonPath());
 
-        int splitElementCount = splitElements.size();
-        if (splitElementCount == 0) {  // Continue the message flow if no matching elements found
-            return true;
-        }
-
-        OMContainer parent = detachedElementContainer.getParent();
-        if (parent == null) {
-            handleException("Error detecting parent element to merge", synCtx);
-        }
-        if (synLog.isTraceOrDebugEnabled()) {
-            synLog.traceOrDebug("Splitting with XPath : " + expression + " resulted in " + splitElementCount +
-                                " elements. Parent node for merging is : " + parent.toString());
-        }
-
-        // iterate through the split elements
-        for (Object element : splitElements) {
-
-            if (!(element instanceof OMNode)) {
-                handleException("Error splitting message with XPath : " + expression +
-                                " - result not an OMNode", synCtx);
+            //Check whether the JSON element expressed by the jsonpath is a valid JsonArray
+            //else throw an exception
+            if (!(iterableChildElements instanceof JsonArray)) {
+                handleException("JSON element expressed by the path "
+                        + ((SynapseJsonPath) expression).getJsonPathExpression()
+                        + " is not a valid JSON array", synCtx);
             }
+            JsonArray iterableJsonArray = iterableChildElements.getAsJsonArray();
             if (synLog.isTraceOrDebugEnabled()) {
-                synLog.traceOrDebug("Submitting " + msgCounter + " of " + splitElementCount +
-                                    " messages for processing in sequentially, in a general loop");
+                synLog.traceOrDebug("Splitting with JSONPath : " + expression + " resulted in " +
+                        iterableJsonArray.size() + " elements.");
             }
+            JsonArray modifiedPayloadArray = new JsonArray();
+            for (JsonElement element : iterableJsonArray) {
+                try {
+                    updateIteratedMessage(synCtx, element);
+                } catch (AxisFault axisFault) {
+                    handleException("Error creating an iterated copy of the message", axisFault, synCtx);
```

Comment:
Would it be better if the error message was changed to "Error updating the stream with iterater element" + element details ?

Reply:
Fixed in https://github.com/wso2/wso2-synapse/pull/1445